File: code/ch10/10.3.3.uniquePaths.py
class Solution:
    visited = dict()

    def uniquePaths(self, m: int, n: int) -> int:
        if (m, n) in self.visited:
            return self.visited[(m, n)]
        if m == 1 or n == 1:
            return 1
        cnt = self.uniquePaths(m - 1, n) + self.uniquePaths(m, n - 1)
        self.visited[(m, n)] = cnt
        return cnt

File: code/ch10/10.3.3.uniquePaths.cpp
#include <utility>
#include <map>
using namespace std;

class Solution {
public:
  map<pair<int, int>, int> visited;
  int uniquePaths(int m, int n) {
    if (visited.count(make_pair(m, n)))
      return visited[make_pair(m, n)];
    if (m == 1 || n == 1) return 1;
    int cnt = uniquePaths(m - 1, n) + \
              uniquePaths(m, n - 1);
    return visited[make_pair(m, n)] = cnt;
  }
};
